internal/testutils/roundrobin/roundrobin.go

```diff
@@ -0,0 +1,213 @@
+// roundrobin-ed across the given backends.
+func CheckRoundRobinRPCs(ctx context.Context, client testgrpc.TestServiceClient, addrs []resolver.Address) error {
+	if err := waitForTrafficToReachBackends(ctx, client, addrs); err != nil {
+		return err
+	}
+
+	// At this point, RPCs are getting successfully executed at the backends
+	// that we care about. To support duplicate addresses (in addrs) and
+	// backends being removed from the list of addresses passed to the
+	// roundrobin LB, we do the following:
+	// 1. Determine the count of RPCs that we expect each of our backends to
+	//    receive per iteration.
+	// 2. Wait until the same pattern repeats a few times, or the context
+	//    deadline expires.
+	wantAddrCount := make(map[string]int)
+	for _, addr := range addrs {
+		wantAddrCount[addr.Addr]++
+	}
+	for ; ctx.Err() == nil; <-time.After(time.Millisecond) {
+		// Perform 3 more iterations.
+		var iterations [][]string
+		for i := 0; i < 3; i++ {
+			iteration := make([]string, len(addrs))
+			for c := 0; c < len(addrs); c++ {
+				var peer peer.Peer
+				if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer)); err != nil {
+					return fmt.Errorf("EmptyCall() = %v, want <nil>", err)
+				}
+				iteration[c] = peer.Addr.String()
+			}
+			iterations = append(iterations, iteration)
+		}
+		// Ensure the the first iteration contains all addresses in addrs.
+		gotAddrCount := make(map[string]int)
+		for _, addr := range iterations[0] {
+			gotAddrCount[addr]++
+		}
+		if diff := cmp.Diff(gotAddrCount, wantAddrCount); diff != "" {
+			logger.Infof("non-roundrobin, got address count in one iteration: %v, want: %v, Diff: %s", gotAddrCount, wantAddrCount, diff)
+			continue
+		}
+		// Ensure all three iterations contain the same addresses.
+		if !cmp.Equal(iterations[0], iterations[1]) || !cmp.Equal(iterations[0], iterations[2]) {
+			logger.Infof("non-roundrobin, first iter: %v, second iter: %v, third iter: %v", iterations[0], iterations[1], iterations[2])
+			continue
+		}
+		return nil
+	}
+	return fmt.Errorf("Timeout when waiting for roundrobin distribution of RPCs across addresses: %v", addrs)
+}
+
+// CheckWeightedRoundRobinRPCs verifies that EmptyCall RPCs on the given
+// ClientConn, connected to a server exposing the test.grpc_testing.TestService,
+// are weighted roundrobin-ed (with randomness) across the given backend
+// addresses.
+//
+// Returns a non-nil error if context deadline expires before RPCs start to get
+// roundrobin-ed across the given backends.
+func CheckWeightedRoundRobinRPCs(ctx context.Context, client testgrpc.TestServiceClient, addrs []resolver.Address) error {
+	if err := waitForTrafficToReachBackends(ctx, client, addrs); err != nil {
+		return err
+	}
+
+	// At this point, RPCs are getting successfully executed at the backends
+	// that we care about. To take the randomness of the WRR into account, we
+	// look for approximate distribution instead of exact.
```

Comment:
This feels really weird to me. I get that we're encapsulating psuedo random numbers in an e2e test with the randomness affecting RPC's, but continuously retrying until it falls within a distribution, especially with how many times this is called just feels off to me. I get that mathmatically perhaps "eventually factors out the randomness and gets it to be within our 5% range." But still "eventually factors out the randomness" is too strong an assertion to make here. I don't know if there is any better way to test this though since pseudo random + e2e. What are your thoughts/suggestions on this @dfawley?

Reply:
I don't know. It feels weird to wrap something like randomness with something like a statistical wrapper around the randomness with, even if it has whatever percentage chance of the randomness falling within a distribution. I chatted with Doug about this, and he mentioned that there's no way around this with regards to randomness. Thus, I am fine with it, even though it feels icky to me.